### PR TITLE
feat(default_ad_api): change to read topic by polling

### PR DESCRIPTION
**Pull request by shtokuda**
## Description
Some callback were replaced with polling subscriber.
I show a list of fixed node in default_ad_api in this PR.
<table>
  <thead>
    <tr>
      <th>Node Name</th>
      <th>Changes Made</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>autoware_state</td>
      <td><b>Yes</b></td>
    </tr>
    <tr>
      <td>localization_conversion</td>
      <td>No</td>
    </tr>
    <tr>
      <td>route_conversion</td>
      <td>No</td>
    </tr>
    <tr>
      <td>diagnostics</td>
      <td>No</td>
    </tr>
    <tr>
      <td>fail_safe</td>
      <td>No</td>
    </tr>
    <tr>
      <td>heart_beat</td>
      <td>No</td>
    </tr>
    <tr>
      <td>interface</td>
      <td>No</td>
    </tr>
    <tr>
      <td>localization</td>
      <td>No</td>
    </tr>
    <tr>
      <td>motion</td>
      <td><b>Yes</b></td>
    </tr>
    <tr>
      <td>operation_mode</td>
      <td>No</td>
    </tr>
    <tr>
      <td>perception</td>
      <td>No</td>
    </tr>
    <tr>
      <td>planning</td>
      <td><b>Yes</b></td>
    </tr>
    <tr>
      <td>routing</td>
      <td><b>Yes</b></td>
    </tr>
    <tr>
      <td>vehicle_door</td>
      <td>No</td>
    </tr>
    <tr>
      <td>vehicle_info</td>
      <td>No</td>
    </tr>
    <tr>
      <td>vehicle</td>
      <td><b>Yes</b></td>
    </tr>
  </tbody>
</table>

</body>
</html>

## Tests performed
The module works as before
I checked published data of following topics on the planning_simulator;

<table>
  <thead>
    <tr>
      <th>Topic Name</th>
      <th>Related Node Name</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>/autoware/state</td>
      <td><b>autoware_state</b></td>
    </tr>
     <tr>
      <td></td>
      <td><b>motion</b></td>
    </tr>
     <tr>
      <td>/api/planning/velocity_factors</td>
      <td rowspan="2"><b>planning</b></td>
    </tr>
     <tr>
      <td>/api/planning/steering_factors</td>
    </tr>
    <tr>
      <td>/api/routing/route</td>
      <td rowspan="2"><b>routing</b></td>
    </tr>
    <tr>
      <td>/api/routing/state</td>
    </tr>
     <tr>
      <td>/api/vehicle/status</td>
      <td rowspan="2"><b>vehicle</b></td>
    </tr>
     <tr>
      <td>/api/vehicle/kinematics</td>
    </tr>
  </tbody>
</table>

</body>
</html>

## Notes for reviewers
The motion node has two subscribers, which subscribe to the 'is_pause' and 'is_start' topics. 
Is it sufficient to update only the on_timer function？

## Effects on system behavior
None

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/

### CODE REVIEW

File: system/default_ad_api/src/compatibility/autoware_state.cpp [hunk 1]
```diff
@@ -79,6 +71,21 @@ void AutowareStateNode::on_shutdown(
...
+  auto routing_msg = routing_sub_->takeData();
+  if (routing_msg) {
+    routing_state_ = *routing_sub_->takeData();
+  }
+
+  auto operation_mode_msg = operation_mode_sub_->takeData();
+  if (operation_mode_msg) {
+    operation_mode_state_ = *operation_mode_sub_->takeData();
+  }
+
```
Comment (shmpwk) [inline system/default_ad_api/src/compatibility/autoware_state.cpp:88]: IMO
Calling `takeData()` twice for each msg is inefficient. Please reuse it.

Comment (shtokuda) [reply system/default_ad_api/src/compatibility/autoware_state.cpp:88]: I agree! I'll make the changes.

File: system/default_ad_api/src/planning.cpp [hunk 2]
```diff
@@ -139,19 +141,19 @@ void PlanningNode::on_timer()
...
+        factor.status = is_stopped ? VelocityFactor::STOPPED : VelocityFactor::APPROACHING;
+      }
     }
-  }
 
-  pub_velocity_factors_->publish(velocity);
-  pub_steering_factors_->publish(steering);
+    pub_velocity_factors_->publish(velocity);
+    pub_steering_factors_->publish(steering);
+  }
```
Comment (shmpwk) [inline system/default_ad_api/src/planning.cpp:156]: What do you intend to execute them if `(trajectory_ && kinematic_state_) ` which originally executed regardless of `trajectory_` and `kinematic_state_`.

File: system/default_ad_api/src/routing.hpp [hunk 3]
```diff
@@ -46,19 +49,27 @@ class RoutingNode : public rclcpp::Node
   Srv<autoware_ad_api::routing::ChangeRoute> srv_change_route_;
   Srv<autoware_ad_api::routing::ClearRoute> srv_clear_route_;
   Sub<planning_interface::RouteState> sub_state_;
-  Sub<planning_interface::LaneletRoute> sub_route_;
+  // Sub<planning_interface::LaneletRoute> sub_route_;
```
Comment (shmpwk) [inline system/default_ad_api/src/routing.hpp:52]:
```suggestion
```

File: system/default_ad_api/src/routing.hpp [hunk 4]
```diff
@@ -46,19 +49,27 @@ class RoutingNode : public rclcpp::Node
...
   Cli<planning_interface::SetWaypointRoute> cli_set_waypoint_route_;
   Cli<planning_interface::SetLaneletRoute> cli_set_lanelet_route_;
   Cli<planning_interface::ClearRoute> cli_clear_route_;
   Cli<system_interface::ChangeOperationMode> cli_operation_mode_;
-  Sub<system_interface::OperationModeState> sub_operation_mode_;
+  // Sub<system_interface::OperationModeState> sub_operation_mode_;
+
+  // std::shared_ptr<tier4_autoware_utils::InterProcessPollingSubscriber<planning_interface::RouteState::Message>>
+  //   state_sub_;
```
Comment (shmpwk) [inline system/default_ad_api/src/routing.hpp:60]:
```suggestion
```

File: system/default_ad_api/src/vehicle.hpp [hunk 5]
```diff
@@ -42,16 +46,33 @@ class VehicleNode : public rclcpp::Node
   rclcpp::CallbackGroup::SharedPtr group_cli_;
```
Comment (shmpwk) [inline system/default_ad_api/src/vehicle.hpp:46]:
```suggestion
```

**Comment by shmpwk** [comment]
@shtokuda 
By the way, how about using https://github.com/autowarefoundation/autoware.universe/blob/0e6528be2af66380b53956c78e75af246b21099c/common/tier4_autoware_utils/include/tier4_autoware_utils/ros/polling_subscriber.hpp ?

File: system/default_ad_api/src/planning.cpp (hunk 2 above)
Comment (shtokuda) [reply system/default_ad_api/src/planning.cpp:156]: Apologies for any confusion.  I misunderstood that published topics in `on_timer` are related to subscribed topics directly. I will revert to the original implementation.

**Comment by shtokuda** [comment]
Sorry, I didn't explain about` create_polling_subscriber`. `create_polling_subscriber` is a wrapper of `InterProcessPollingSubscriber` to set arguments from `component_interface_specs`. Here, we use the wrapper to improve readability.

**Comment by shmpwk** [comment]
> The motion node has two subscribers, which subscribe to the 'is_pause' and 'is_start' topics.
Is it sufficient to update only the on_timer function？

I overlooked this :pray: 
I agree to take `is_pause` and `is_start` in timer callback.

**Comment by shmpwk** [comment]
@isamu-takagi 
Could you review this? I wonder we may break the way of default_ad_api.

**Comment by isamu-takagi** [comment]
@shmpwk There are two subscriber wrappers that conflict. Can you wait while I work on a solution for component_interface_utils? It would also be useful to have an implementation of polling subscription using a free function.

**Comment by isamu-takagi** [comment]
@shmpwk I created PR without polling subscriber wrapper. Could you review it? 
https://github.com/autowarefoundation/autoware.universe/pull/7588

**Comment by isamu-takagi** [comment]
Close because https://github.com/autowarefoundation/autoware.universe/pull/7588 has taken over.

**Pull request closed without merging**